### ISSUE TITLE
SyntaxTree diffing algorithm performs very poorly (leading to Fix-All scenarios being broken).

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#17498]
The code-action fix-all engine works by having each registered action perform it's fix, getting the diff of that fix against the original document, and then trying to merge together all those diffs.

As such, for fix-all to work properly, the diffs produced by the diffing algorithm need to be small and reasonably accurate.  Otherwise, if the diffs are large they will end up looking like overlapping changes, and the fix all engine will not merge them in (leading to incomplete 'fix-alls').

The way we get diffs is inside Document.GetTextChangesAsync.  This eventually calls:

```c#
                    if (this.SupportsSyntaxTree)
                    {
                        var tree = await this.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
                        var oldTree = await oldDocument.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);

                        return tree.GetChanges(oldTree);
                    }
```

Ostensibly, this should get 'good' results as we should be able to get minimal diffs by accurately comparing the syntax trees.  Unfortunately, for reasons i have not investigated heavily, the syntax tree diffing algorithm performs quite poorly in very simple cases.  For example, in a very large file that only contains the following diff:

![image](https://cloud.githubusercontent.com/assets/4564579/23478400/208d438e-fe76-11e6-8f44-aca729b5fdd7.png)

The syntax-tree differ produces 14 different changes, one of which is 800 characters long.  These changes then conflict with other changes made, and we end up not actually 'fixing all'.

![image](https://cloud.githubusercontent.com/assets/4564579/23478533/810d8e26-fe76-11e6-9222-7bcc8a780262.png)

I'm attaching the two files i'm using to diff, as well as the tiny sample program to reveal the issue.

**Comment by CyrusNajmabadi** [comment]
Here are the files to diff.

[Files.zip](https://github.com/dotnet/roslyn/files/811887/Files.zip)

Issue can be reproduced with the following code:

```c#
using System;
using System.IO;
using Microsoft.CodeAnalysis.VisualBasic;

class Program
{
    static void Main(string[] args)
    {
        var tree1 = VisualBasicSyntaxTree.ParseText(File.ReadAllText(@"c:\temp\before.txt"));
        var tree2 = VisualBasicSyntaxTree.ParseText(File.ReadAllText(@"c:\temp\after.txt"));

        var changes = tree1.GetChanges(tree2);
        Console.WriteLine(changes.Count);
    }
}
```

**Comment by mattwar** [comment]
The repro test is bad because it is based on two separately parsed files.  The syntax differ is based on recognizing the same green node instances. Nodes that are just similar have a much smaller correlation.

**Comment by CyrusNajmabadi** [comment]
That's a serious problem then :)  As this is what ends up getting used during Fix-All and this is how Document.GetTextChangesAsync works.  Perhaps we should only do the tree diffing if we can verify that they came from trees that have an incremental parsing relationship?   i.e. any time we create a new fresh tree, we give it an ID. Then, if we incrementally parse, we use that same ID.  Then, we only do syntax diffing if the incremental IDs are the same.  Otherwise we just do text diffing.

**Comment by tmat** [comment]
The core diffing algorithm doesn't rely on reusing syntax nodes. It should be possible to change the alg customization to do value comparison of nodes.

**Comment by mattwar** [comment]
When you call GetChanges on a SourceText that did not derive from the older source text argument, you just get one change referencing the entire span. There is no attempt to actual diff the text. Likewise, with document.GetChanges, it uses the SourceText's changes if they exist and are not just the entire document, otherwise it attempts to make a quick best effort with the syntax trees. The assumption is that either the text or the tree's were modified by a few local edits, and not wholesale replaced.  Typically this works very well.  It was not designed to be as precise in all cases as the fix all provider is requiring. This has come up in a few bug reports in the past. I pushed back each time there too, but eventually found a small change that would improve the results without tanking the speed.

**Comment by mattwar** [comment]
It is possible that we could switch to use the tree diffing code that @tmat added, but it may bust the perf tests.

**Comment by CyrusNajmabadi** [comment]
>  This has come up in a few bug reports in the past. I pushed back each time there too, but eventually found a small change that would improve the results without tanking the speed.

The problem is then that we simply cannot use this API for FixAll.  I personally don't think that's a great situation to be in.

>  The assumption is that either the text or the tree's were modified by a few local edits, and not wholesale replaced. 

This isn't borne out by the code.  You can have a few local edits (which is the case i have above), and still get terrible diffs.  This is because the code assumes that it can get back to the original SourceTexts for each Document and then get the changes off of that.  In a scenario like FixAll, we may have dumped the SourceText, leading us to then go to hte bad tree-diff.

There is no guarantee that a small text change leads to a Document where both SourceTexts are alive.  As such, the diff can do terribly even with *small* changes.  this is precisely the issue that made me file this bug.  We only made a tiny change to the document, but Document.GetTextChangesAsync returned an absolutely *terrible* set of results.

**Comment by tmat** [comment]
Oh, I thought the tree diff code is already being used :).
Yes, the alg is O(N^2). EnC does multi-level diffing: declaration level, method body level, lambda level. In each case only nodes on the level (declarations, statements in the method, statements in the lambda, respectively) are being considered.

**Comment by CyrusNajmabadi** [comment]
> It is possible that we could switch to use the tree diffing code that @tmat added, but it may bust the perf tests.

Before doing that, i would propose the following:

1. We only do tree diffing if we can validate (in some manner) that these trees are incrementally related.  I outlined a way we could do that above.
2. If we cannot do tree diffing we fall back to SourceText diffing. 
3. If that immediately returns a single 'full span changed' result we then try to import the IDocumentDifferenceService and we request that it do the diff.

This way you always get a *decent* result, but we try to keep the perf as good as possible.  namely, if we have real trees, with shared greed nodes, we can get a good small diff.  Otherwise, if we have SourceTexts with stored TextChanges between them, we can return that.  Otherwise, we fall back to a real diff that can give small results.

Today we have a bad situation, we can end up with a tree diff that produces bad results, and a SourceText change that can result in saying "the entire document changed".  These always end up being bad for the consumer that wants some sort of reasonable diff.  The above algorithm sidesteps the bad diffs, and produces a reasonable diff when we don't have anything smarter to do.

**Comment by mattwar** [comment]
@CyrusNajmabadi I thought you said you were getting a bad result even when they are incrementally related.  Maybe we should examine that problem first.

**Comment by CyrusNajmabadi** [comment]
"even when they are incrementally related".  They are incrementally related, but that does not mean that the old text or old tree are available.  As such:

```c#
 var tree = await this.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
                        var oldTree = await oldDocument.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);

                        return tree.GetChanges(oldTree);
```

may test two completely parsed trees against each other.

**Comment by CyrusNajmabadi** [comment]
@jinujoseph Can this be assigned to someone on Roslyn.  Perhaps @tmat as he has SyntaxDiffing experience through EnC?

**Comment by CyrusNajmabadi** [comment]
We got another report of this from @gix

His repro involves replacing a string literal in argument position, with a series of arguments.  i.e. his starting code is `new Guid("5636571C-2B61-442E-BD6B-199A7BFA1B3B")` and his fixer updates that to `new Guid(0x5636571C, 0x2B61, 0x442E, 0xBD, 0x6B, 0x19, 0x9A, 0x7B, 0xFA, 0x1B, 0x3B)`.   This is a 'large enough' difference that the differ goes completes off the rails.

The code for his analyzer is:

<details>
```c#
namespace GuidFromStringAnalysis
{
    using System;
    using System.Collections.Immutable;
    using Microsoft.CodeAnalysis;
    using Microsoft.CodeAnalysis.CSharp;
    using Microsoft.CodeAnalysis.CSharp.Syntax;
    using Microsoft.CodeAnalysis.Diagnostics;

    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    internal class GuidFromStringAnalyzer : DiagnosticAnalyzer
    {
        public const string DiagnosticId = "GuidFromString";

        private const string Category = "AnalyzerPerformance";
        const string analyzerTitle = "Title";

        //private static readonly LocalizableString Title =
        //    new LocalizableResourceString(
        //        nameof(Resources.AnalyzerTitle),
        //        Resources.ResourceManager,
        //        typeof(Resources));

        //private static readonly LocalizableString MessageFormat =
        //    new LocalizableResourceString(
        //        nameof(Resources.AnalyzerMessageFormat),
        //        Resources.ResourceManager,
        //        typeof(Resources));

        //private static readonly LocalizableString Description =
        //    new LocalizableResourceString(
        //        nameof(Resources.AnalyzerDescription),
        //        Resources.ResourceManager,
        //        typeof(Resources));

        private static readonly DiagnosticDescriptor Rule =
            new DiagnosticDescriptor(
                DiagnosticId, "Guid title", "Guid message", Category,
                DiagnosticSeverity.Info, isEnabledByDefault: true,
                description: "Guid description");

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);

        public override void Initialize(AnalysisContext context)
        {
            context.RegisterSyntaxNodeAction(AnalyzeObjectCreation, SyntaxKind.ObjectCreationExpression);
        }

        private static void AnalyzeObjectCreation(SyntaxNodeAnalysisContext context)
        {
            var creationExpr = (ObjectCreationExpressionSyntax)context.Node;

            var symbolInfo = context.SemanticModel.GetSymbolInfo(creationExpr.Type);
            var typeSymbol = symbolInfo.Symbol as INamedTypeSymbol;
            if (typeSymbol == null || typeSymbol.ToString() != typeof(Guid).FullName)
                return;

            var arguments = creationExpr.ArgumentList.Arguments;
            if (arguments.Count == 1 && arguments[0].Expression.Kind() == SyntaxKind.StringLiteralExpression)
            {
                var diagnostic = Diagnostic.Create(Rule, arguments[0].GetLocation());
                context.ReportDiagnostic(diagnostic);
            }
        }
    }
}
```

```c#
namespace GuidFromStringAnalysis
{
    using System;
    using System.Collections.Immutable;
    using System.Composition;
    using System.Linq;
    using System.Threading;
    using System.Threading.Tasks;
    using Microsoft.CodeAnalysis;
    using Microsoft.CodeAnalysis.CodeActions;
    using Microsoft.CodeAnalysis.CodeFixes;
    using Microsoft.CodeAnalysis.CSharp;
    using Microsoft.CodeAnalysis.CSharp.Syntax;
    using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;

    [ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(GuidFromStringCodeFixProvider))]
    [Shared]
    internal class GuidFromStringCodeFixProvider : CodeFixProvider
    {
        private const string Title = "Construct from integers";

        public sealed override ImmutableArray<string> FixableDiagnosticIds =>
            ImmutableArray.Create(GuidFromStringAnalyzer.DiagnosticId);

        public sealed override FixAllProvider GetFixAllProvider()
        {
            return WellKnownFixAllProviders.BatchFixer;
        }

        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
        {
            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);

            var diagnostic = context.Diagnostics.First();
            var diagnosticSpan = diagnostic.Location.SourceSpan;

            var expr = root.FindToken(diagnosticSpan.Start).Parent.AncestorsAndSelf()
                .OfType<ObjectCreationExpressionSyntax>().First();

            context.RegisterCodeFix(
                CodeAction.Create(
                    title: Title,
                    createChangedDocument: c => ConstructGuidFromIntegersAsync(context.Document, expr, c),
                    equivalenceKey: Title),
                diagnostic);
        }

        private async Task<Document> ConstructGuidFromIntegersAsync(
            Document document, ObjectCreationExpressionSyntax creationExpr, CancellationToken cancellationToken)
        {
            var argumentList = creationExpr.ArgumentList;
            var argument = argumentList.Arguments[0];
            var literal = (LiteralExpressionSyntax)argument.Expression;
            var guidStr = literal.Token.ValueText;
            if (!Guid.TryParse(guidStr, out var guid))
                return document;

            var bytes = guid.ToByteArray();
            var data1 = BitConverter.ToUInt32(bytes, 0);
            var data2 = BitConverter.ToUInt16(bytes, 4);
            var data3 = BitConverter.ToUInt16(bytes, 6);

            var newArgList = creationExpr.ArgumentList.WithArguments(
                SeparatedList(new [] {
                    // Add any trivia of the original argument to preserve whitespace.
                    Argument(PaddedHexLiteral(data1)).WithLeadingTrivia(argument.GetLeadingTrivia()),
                    Argument(PaddedHexLiteral(data2)),
                    Argument(PaddedHexLiteral(data3)),
                    Argument(PaddedHexLiteral(bytes[8])),
                    Argument(PaddedHexLiteral(bytes[9])),
                    Argument(PaddedHexLiteral(bytes[10])),
                    Argument(PaddedHexLiteral(bytes[11])),
                    Argument(PaddedHexLiteral(bytes[12])),
                    Argument(PaddedHexLiteral(bytes[13])),
                    Argument(PaddedHexLiteral(bytes[14])),
                    Argument(PaddedHexLiteral(bytes[15]).WithTrailingTrivia(argument.GetTrailingTrivia()))
                }));

            var oldRoot = await document.GetSyntaxRootAsync(cancellationToken);
            var newRoot = oldRoot.ReplaceNode(argumentList, newArgList);

            return document.WithSyntaxRoot(newRoot);
        }

        private static LiteralExpressionSyntax PaddedHexLiteral(byte value)
        {
            return LiteralExpression(
                SyntaxKind.NumericLiteralExpression,
                Literal("0x" + value.ToString("X2"), value));
        }

        private static LiteralExpressionSyntax PaddedHexLiteral(ushort value)
        {
            return LiteralExpression(
                SyntaxKind.NumericLiteralExpression,
                Literal("0x" + value.ToString("X4"), value));
        }

        private static LiteralExpressionSyntax PaddedHexLiteral(uint value)
        {
            return LiteralExpression(
                SyntaxKind.NumericLiteralExpression,
                Literal("0x" + value.ToString("X8"), value));
        }
    }
}
```

TestFile:
```c#
namespace Namespace1
{
    using System;

    public class Class1
    {
        private void F0()
        {
            var id = new Guid("5636571C-2B61-442E-BD6B-199A7BFA1B3B");
        }

        private void F1()
        {
        }

        private void F2()
        {
        }

        private static readonly Guid f1;
        private static readonly Guid f2;

        private void G1()
        {
        }

        private void G2()
        {
        }

        private void G3()
        {
        }

        private void G4()
        {
        }

        private void G5()
        {
        }

        private void G6()
        {
        }

        private void G7()
        {
        }

        private void G8()
        {
        }

        private void G9()
        {
        }

        private void G10()
        {
        }
    }
}
```

**Comment by CyrusNajmabadi** [comment]
Note: i've debugged through this, and the core problem is in SyntaxDiffer.FindBestMatch.  It looks like it's limit for finding matches is so small that even basic changes like this totally bork things.  For example, if i increased SyntaxDiffer.MaxSearchLength to 24, then the diff turns out fine.

Basically, when diffing, we have hte `)` in `new Guid("...")`.  But we can't find that same `)` in `new Guid(num, num, ..., num)`.  At this point, everything goes bad.  The differ starts replacing each successive token in the source and dest and ends up feeling like the entire remainder of the file has been changed.

**Comment by CyrusNajmabadi** [comment]
One suggestion i would have would be to actually keep track of push-pop regions as we diff hte trees.  The main problem here is that even when we push `num, num, ..., num` into our diffing stack, we completely lose that we got in there by 'push'ing in from `(...)`.  If we kept track of that, we could then use those pop-points as potential re-sync points, allowing us to get back on track and minimize the damage we've done.
